src/sys/quota.rs

```diff
@@ -1,78 +1,235 @@
+        QFMT_VFS_V1,
+    }
+}
 
-    impl QuotaCmd {
-        pub fn as_int(&self) -> c_int {
-            ((self.0 << 8) | (self.1 & 0x00ff)) as c_int
+libc_bitflags!(
+    /// Indicates the quota fields that are valid to read from.
+    #[derive(Default)]
+    pub struct QuotaValidFlags: u32 {
+        /// The block hard & soft limit fields.
+        QIF_BLIMITS;
+        /// The current space field.
+        QIF_SPACE;
+        /// The inode hard & soft limit fields.
+        QIF_ILIMITS;
+        /// The current inodes field.
+        QIF_INODES;
+        /// The disk use time limit field.
+        QIF_BTIME;
+        /// The file quote time limit field.
+        QIF_ITIME;
+        /// All block & inode limits.
+        QIF_LIMITS;
+        /// The space & inodes usage fields.
+        QIF_USAGE;
+        /// The time limit fields.
+        QIF_TIMES;
+        /// All fields.
+        QIF_ALL;
+    }
+);
+
+/// Wrapper type for `if_dqblk`
+// FIXME: Change to repr(transparent)
+#[repr(C)]
+#[derive(Clone, Copy)]
+pub struct Dqblk(libc::dqblk);
+
+impl Default for Dqblk {
+    fn default() -> Dqblk {
+        Dqblk(libc::dqblk {
+            dqb_bhardlimit: 0,
+            dqb_bsoftlimit: 0,
+            dqb_curspace: 0,
+            dqb_ihardlimit: 0,
+            dqb_isoftlimit: 0,
+            dqb_curinodes: 0,
+            dqb_btime: 0,
+            dqb_itime: 0,
+            dqb_valid: 0,
+        })
+    }
+}
+
+impl Dqblk {
+    /// The absolute limit on disk quota blocks allocated.
+    pub fn blocks_hard_limit(&self) -> Option<u64> {
+        let valid_fields = QuotaValidFlags::from_bits_truncate(self.0.dqb_valid);
+        if valid_fields.contains(QIF_BLIMITS) {
+            Some(self.0.dqb_bhardlimit)
+        } else {
+            None
         }
     }
 
-    // linux quota version >= 2
-    pub const Q_SYNC:	QuotaSubCmd = 0x800001;
-    pub const Q_QUOTAON:	QuotaSubCmd = 0x800002;
-    pub const Q_QUOTAOFF:	QuotaSubCmd = 0x800003;
-    pub const Q_GETFMT:	QuotaSubCmd = 0x800004;
-    pub const Q_GETINFO:	QuotaSubCmd = 0x800005;
-    pub const Q_SETINFO:	QuotaSubCmd = 0x800006;
-    pub const Q_GETQUOTA:	QuotaSubCmd = 0x800007;
-    pub const Q_SETQUOTA:	QuotaSubCmd = 0x800008;
-
-    pub type QuotaType = c_int;
-
-    pub const USRQUOTA:	QuotaType = 0;
-    pub const GRPQUOTA:	QuotaType = 1;
-
-    pub type QuotaFmt = c_int;
-
-    pub const QFMT_VFS_OLD:	QuotaFmt = 1;
-    pub const QFMT_VFS_V0:	QuotaFmt = 2;
-    pub const QFMT_VFS_V1:  QuotaFmt = 4;
-
-    libc_bitflags!(
-        #[derive(Default)]
-        pub struct QuotaValidFlags: u32 {
-            QIF_BLIMITS;
-            QIF_SPACE;
-            QIF_ILIMITS;
-            QIF_INODES;
-            QIF_BTIME;
-            QIF_ITIME;
-            QIF_LIMITS;
-            QIF_USAGE;
-            QIF_TIMES;
-            QIF_ALL;
+    /// Set the absolute limit on disk quota blocks allocated.
+    pub fn set_blocks_hard_limit(&mut self, limit: u64) {
```

Comment:
Should this method also set `QIF_BLIMITS`?

Reply:
I don't think so. If anything, it should unset it. This API basically has two modes: reading the current state and modifying the existing state. The `QIF_` values are used to indicate which values are set in both cases. So in case you read, modify, then write, we actually want to completely ignore the value set in the struct, as it might be left-over from a read, and instead force the user to specify it explicitly for this specific write operation. That way there's no confusion.

I think we could simplify the API by doing something like what you suggest but actually use separate fields to track reading and writing and then setting the `writing_valid_fields` field to the proper flags when any of the appropriate `set_*` functions are called might work. Thoughts?